### PR TITLE
CI: add the first pipeline to build the application and run basic tests

### LINKED ISSUE #6: Look into Github Actions for CMake projects
Github has a suggested action about CMake projects, so it would be nice if someone could look into how to setup it, what can be done with it, how it can help the project...

I think having an auto-build executable available to download  after each new release would be nice for example, but maybe we can do more fancy stuff, such as auto checking tests (if we write tests...) to help have faster PR, etc...

### PR DESCRIPTION
**Pull request by amtoine**
This PR should close #6.
Requires
- #2 to be merged

---
This PR adds a slightly modified default action from _GitHub_ for `CMake`-based projects :+1: 

I did run the pipeline on my personal fork, which resulted in [this action log](https://github.com/amtoine/wiresmash/actions/runs/3670185074/jobs/6204543791) on the tip of the branch of #2 :+1: 
this means that the tests will pass when the PR is ready :muscle: 

## NOTES
- this pipeline won't work on `windows` hence the single `ubuntu-latest` in the matrix of OSes :thinking: 
- i've let the comments from the default configuration file, just for reference :+1: 
- there are not tests to run, so they pass :ok_hand:

### CODE REVIEW

**Comment by gbrivady** [comment]
Is it fine with you if we wait for #2 to be done before merging? I know it is not necessary, but I would prefer if we keep the "order" of PR coherent with the milestones assigned to them

**Comment by amtoine** [comment]
> Is it fine with you if we wait for #2 to be done before merging? I know it is not necessary, but I would prefer if we keep the "order" of PR coherent with the milestones assigned to them
>
fine it is :wink: 

maybe still do the review such that we can merge #16 right after #2? :yum:

File: .github/workflows/build.yml
Comment (gbrivady) [inline .github/workflows/build.yml:4]: Just to be sure : this is on which branch the test will run, when Github detects a push/a pull request?

**Comment by gbrivady** [comment]
Just one question, and then I trust that you tested it well, we will see if it does not work after the pull anyway

File: .github/workflows/build.yml
Comment (amtoine) [reply .github/workflows/build.yml:4]: let's explain everything, shall we? :yum: 

- the `on` field specifies the events on which the pipeline is automatically triggered
- nowadays, there is only the `main` branch and it is push-protected, meaning that we can not manually push to it and the settings require to go through a review process and a PR :ok_hand: 
- the `on.pull_request.branches` field gives the list of all the branches that will trigger a CI job when a PR is opened on them
for now, it's obviously only the `main` branch
- the `on.push.branches` is quite the same but for direct pushes to the branches in the list, i.e. the `main` branch only for now

## an example to wrap this up!
- i have a bunch of commit on the `main` branch
- i try to push them on `iScsc/wiresmash/main`, but this is refused thanks to the branch protection
- so i create a `feature` branch, push to either `iSCsc/wiresmash` or a fork of mine
- i open a PR onto `iScsc/wiresmash/main`
- as this is part of `on.pull_request.branches` a CI job is triggered in the PR body
this job can, when properly configured, block a PR until it does not fail anymore, very handy to block a failing source base :ok_hand: 
- the review process goes on
- we merge the PR -> **this is a push on the `main` branch** so, as it is part of `on.push.branches`, this triggers a CI job on the `main` branch, appearing on the front page of the repo, with a :red_circle:, a :orange_circle: or a :green_circle:  depending on the status of the last CI job :star_struck: 

## about the new branches
yes, depending on the new branches and their meaning and use, we will add them somewhere in the `on` field and in the branch protection rules list :+1:

**Comment by amtoine** [comment]
> Just one question, and then I trust that you tested it well, we will see if it does not work after the pull anyway; so good for me
>
you do not even have to trust me :smirk: 

here's what we could do
- wait for #2 to be merged into the `main` branch
- merge the PR branch here into the `main` branch to get the new source base that should pass the tests of the pipeline
- let the tests run just as in the [job above](https://github.com/iScsc/wiresmash/actions/runs/3670221824/jobs/6204609052)
- see if they pass the tests
- finally merge this PR

File: .github/workflows/build.yml
Comment (gbrivady) [reply .github/workflows/build.yml:4]: Ty for the explanation, I got the gist of it

Comment (amtoine) [reply .github/workflows/build.yml:4]: greeeat :relieved: :muscle:

**Review by ctmbl — APPROVED** [review]
I like that! everything's fine to me :+1:

**Comment by amtoine** [comment]
> I like that! everything's fine to me +1
>
great :relieved: 

then @gbrivady, you can review this once #1 is merged :wink:

**Comment by amtoine** [comment]
@gbrivady, #2 has been merged, i think we can move on with this PR :wink: 

first, we need to make sure the source compiles on the latest `main` :+1:

**Comment by amtoine** [comment]
looks it works fine on the latest `main`, after #2 :yum:

**Comment by gbrivady** [comment]
Works, so good for me :+1:

**Pull request merged**
